src/Lib.P2P/Multiplex/Substream.cs

```diff
@@ -0,0 +1,240 @@
+#region LICENSE
+
+/**
+* Copyright (c) 2019 Catalyst Network
+*
+* This file is part of Catalyst.Node <https://github.com/catalyst-network/Catalyst.Node>
+*
+* Catalyst.Node is free software: you can redistribute it and/or modify
+* it under the terms of the GNU General Public License as published by
+* the Free Software Foundation, either version 2 of the License, or
+* (at your option) any later version.
+*
+* Catalyst.Node is distributed in the hope that it will be useful,
+* but WITHOUT ANY WARRANTY; without even the implied warranty of
+* MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+* GNU General Public License for more details.
+*
+* You should have received a copy of the GNU General Public License
+* along with Catalyst.Node. If not, see <https://www.gnu.org/licenses/>.
+*/
+
+#endregion
+
+using System;
+using System.IO;
+using System.Threading;
+using System.Threading.Tasks;
+using System.Threading.Tasks.Dataflow;
+using MultiFormats;
+
+namespace Lib.P2P.Multiplex
+{
+    /// <summary>
+    ///   A duplex substream used by the <see cref="Muxer"/>.
+    /// </summary>
+    /// <remarks>
+    ///   Reading of data waits on the Muxer calling <see cref="AddData(byte[])"/>.
+    ///   <see cref="NoMoreData"/> is used to signal the end of stream.
+    ///   <para>
+    ///   Writing data is buffered until <see cref="FlushAsync(CancellationToken)"/> is
+    ///   called.
+    ///   </para>
+    /// </remarks>
+    public class Substream : Stream
+    {
+        private BufferBlock<byte[]> inBlocks = new BufferBlock<byte[]>();
+        private byte[] inBlock;
+        private int inBlockOffset;
+        private bool eos;
+
+        private Stream outStream = new MemoryStream();
+
+        /// <summary>
+        ///   The type of message of sent to the other side.
+        /// </summary>
+        /// <value>
+        ///   Either <see cref="PacketType.MessageInitiator"/> or <see cref="PacketType.MessageReceiver"/>.
+        ///   Defaults to <see cref="PacketType.MessageReceiver"/>.
+        /// </value>
+        public PacketType SentMessageType = PacketType.MessageReceiver;
+
+        /// <summary>
+        ///   The stream identifier.
+        /// </summary>
+        /// <value>
+        ///   The session initiator allocates odd IDs and the session receiver allocates even IDs.
+        /// </value>
+        public long Id;
```

Comment:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Make this field 'private' and encapsulate it in a 'public' property.](https://app.codacy.com/manual/catalyst-network/Catalyst.Framework/pullRequest?prid=4832783)